src/com/amazon/ion/impl/IonCursorBinary.java

```diff
@@ -0,0 +1,721 @@
+            return true;
+        } else if (parent.endIndex < peekIndex) {
+            throw new IonException("Contained values overflowed the parent container length.");
+        } else if (parent.typeId.type == IonType.STRUCT) {
+            if (minorVersion == 0) {
+                byte b = buffer[(int) peekIndex++];
+                if (b < 0) {
+                    fieldSid = (b & LOWER_SEVEN_BITS_BITMASK);
+                } else {
+                    fieldSid = (int) readVarUInt_1_0(b);
+                }
+            } else {
+                readFieldName_1_1();
+            }
+        }
+        valuePreHeaderIndex = peekIndex;
+        return false;
+    }
+
+    /**
+     * Reports the total number of bytes consumed from the stream since the last report, up to the current `peekIndex`.
+     */
+    private void reportConsumedData() {
+        long totalNumberOfBytesRead = getTotalOffset() + (peekIndex - valuePreHeaderIndex);
+        dataHandler.onData((int) (totalNumberOfBytesRead - lastReportedByteTotal));
+        lastReportedByteTotal = totalNumberOfBytesRead;
+    }
+
+    /**
+     * Advances to the next token, seeking past the previous value if necessary. After return `event` will convey
+     * the result (e.g. START_SCALAR, END_CONTAINER)
+     * @return false if the next token was an Ion version marker or NOP pad; otherwise, true. If false, this method
+     *  should be called again to advance to the following value.
+     */
+    private boolean nextToken() {
+        if (peekIndex < valueMarker.endIndex) {
+            peekIndex = valueMarker.endIndex;
+        } else if (valueTid != null && valueTid.isDelimited) {
+            seekPastDelimitedContainer_1_1();
+        }
+        valueTid = null;
+        if (dataHandler != null) {
+            reportConsumedData();
+        }
+        if (peekIndex >= limit) {
+            setCheckpointBeforeUnannotatedTypeId();
+            if (parent != null && parent.endIndex == peekIndex) {
+                event = Event.END_CONTAINER;
+            }
+            return true;
+        }
+        reset();
+        int b;
+        if (parent == null) { // Depth 0
+            valuePreHeaderIndex = peekIndex;
+            b = buffer[(int)(peekIndex++)] & SINGLE_BYTE_MASK;
+            if (b == IVM_START_BYTE) {
+                readIvm();
+                return false;
+            }
+        } else {
+            if (nextContainedToken()) {
+                return true;
+            }
+            b = buffer[(int)(peekIndex++)] & SINGLE_BYTE_MASK;
+        }
+        if (readHeader(b, false, valueMarker)) {
+            valueTid = valueMarker.typeId;
+            return true;
+        }
+        valueTid = valueMarker.typeId;
+        return false;
+    }
+
+    @Override
+    public Event nextValue() {
+        event = Event.NEEDS_DATA;
+        while (true) {
+            if (nextToken()) {
+                break;
+            }
+        }
```

Comment:
I assume this is faster than
```
while (nextToken()) { }
```
(obviously we'd need to invert `nextToken`)

Reply:
Nope, will clean up.

Reply:
Addressed in https://github.com/amazon-ion/ion-java/pull/538/files